core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java

```diff
@@ -214,71 +224,89 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
                         try {
                             channel.sendResponse(TransportResponse.Empty.INSTANCE);
                         } catch (Exception channelException) {
-                            logger.warn("{} failed to send response while failing shard [{}]", channelException, request.shardRouting.shardId(), request.shardRouting);
+                            logger.warn("{} failed to send response while failing shard [{}]", channelException, request.shardId, request);
                         }
                     }
                 }
             );
         }
     }
 
-    static class ShardFailedClusterStateTaskExecutor implements ClusterStateTaskExecutor<ShardRoutingEntry> {
+    public static class ShardFailedClusterStateTaskExecutor implements ClusterStateTaskExecutor<ShardEntry> {
         private final AllocationService allocationService;
         private final RoutingService routingService;
         private final ESLogger logger;
 
-        ShardFailedClusterStateTaskExecutor(AllocationService allocationService, RoutingService routingService, ESLogger logger) {
+        public ShardFailedClusterStateTaskExecutor(AllocationService allocationService, RoutingService routingService, ESLogger logger) {
             this.allocationService = allocationService;
             this.routingService = routingService;
             this.logger = logger;
         }
 
         @Override
-        public String describeTasks(List<ShardRoutingEntry> tasks) {
-            return tasks.stream().map(entry -> entry.getShardRouting().toString()).reduce((s1, s2) -> s1 + ", " + s2).orElse("");
-        }
-
-        @Override
-        public BatchResult<ShardRoutingEntry> execute(ClusterState currentState, List<ShardRoutingEntry> tasks) throws Exception {
-            BatchResult.Builder<ShardRoutingEntry> batchResultBuilder = BatchResult.builder();
-
-            // partition tasks into those that correspond to shards
-            // that exist versus do not exist
-            Map<ValidationResult, List<ShardRoutingEntry>> partition =
-                tasks.stream().collect(Collectors.groupingBy(task -> validateTask(currentState, task)));
+        public BatchResult<ShardEntry> execute(ClusterState currentState, List<ShardEntry> tasks) throws Exception {
+            BatchResult.Builder<ShardEntry> batchResultBuilder = BatchResult.builder();
+            List<ShardEntry> validTasks = new ArrayList<>();
```

Comment:
maybe call this pendingTasks or resolvedTasks? I got a bit confused by the valid notion - some of the tasks are marked as successful but are not valid :)